File: py/prob_6.py
class Solution:
    def convert(self, s, numRows):
        """
        :type s: str
        :type numRows: int
        :rtype: str
        """
        if numRows > 1:
            wordMatrix = []
            
            while s:
                vertical = s[:numRows]
                diagonal = s[numRows:2*(numRows-1)][::-1]
                diagonal = (numRows - len(diagonal) - 1) * " " + diagonal
                wordMatrix.append(vertical) if vertical else None
                wordMatrix.append(diagonal) if diagonal else None
                s = s[2*(numRows-1):]
                
            ans = ""
            for row_i in range(numRows):
                for word in wordMatrix:
                    ans += word[row_i:row_i+1]
            ans = ans.replace(" ", "")
            return ans
        else:
            return s

File: js/prob_6.js
/**
 * @param {string} s
 * @param {number} numRows
 * @return {string}
 */
const convert = (s, numRows) => {
    if (numRows > 1) {
        let wordMatrix = []
        
        while (s) {
            let vertical = s.slice(0, numRows)
            let diagonal = s.slice(numRows, 2*(numRows-1))
                            .split("").reverse().join("")
            diagonal = " ".repeat(numRows - diagonal.length - 1).concat(diagonal)
            vertical ? wordMatrix.push(vertical) : null
            vertical ? wordMatrix.push(diagonal) : null
            s = s.slice(2*(numRows-1))
        }
        let ans = ""
        for (let row_i=0; row_i < numRows; row_i++) {
            for (let word of wordMatrix) {
                ans = ans.concat(word.slice(row_i, row_i+1))
            }
        }
        ans = ans.replace(/ /g, "")
        return ans
    }
    else {
        return s
    }
}